Provide a default TimeZone to SqueezeOS based players

mw9 opened the pull request:
This proposed change implements an `http` endpoint `/tz` that provides the TimeZone in which the local LMS server operates.

It replaces a previous change proposal that had been made against the Community Firmware plugin. That can be seen [here]( https://github.com/ralph-irving/plugin-CommunityFirmware/pull/8). The substantive change to that proposal is that a simple web endpoint is now used, rather than using the Slim Request/Query mechanism.

The endpoint returns an Olson formatted TimeZone string with a 200, 'ok', response code. If the server fails to obtain a valid TimeZone it will simply return a 404, 'not found',  response.

A companion change to SqueezeOS/Squeezeplay has been prepared to enable it to make use of this service. [This PR](https://github.com/ralph-irving/squeezeos-squeezeplay/pull/15) refers.

The purpose of this endpoint is to allow SqueezeOS based devices to obtain a default TimeZone if their own TimeZone has not been initialized. This would arise following a “factory reset” of the device, or on first set up. It replaces the service that used to be provided by MySqueezebox.com.

The request handler obtains a best guess of its local TimeZone by making an API call to `https://stats.lms-community.org/api/time`, which has been recently implemented. This has the advantage of being platform neutral, and removes the need to “sniff out” a TimeZone from the OS on which LMS runs.

After retrieving the result of the API call, the handler carries out a number of validation checks to verify that the result is properly formatted, and that the TimeZone string basically complies with the naming conventions used within the IANA Time Zone Database. If verification fails, the query will return ~~the empty string~~ a 404 error.

The naming conventions may be found in the Timezone identifiers section of the document Theory and pragmatics of the tz code and data. Refer [here](https://ftp.iana.org/tz/tzdb-2022b/theory.html#naming) for details.

The proposed change has been tested locally on a MacOS based system, and appears to operate as expected.

mw9 committed:
Provide a default TimeZone to SqueezeOS based players

This change implements a new '/tz' HTTP endpoint whose purpose is to provide
SqueezeOS based players with a default TimeZone. The TimeZone is obtained from
'https://stats.lms-community.org/api/time', and will be based on the location
of the LMS server machine.

LMS responds by returning an Olson formatted TimeZone string, with a '200'
(RC_OK) response code. If the server fails to obtain a valid TimeZone it
returns a '404' (RC_NOT_FOUND) response.

A corresponding change has been made to the Community Firmware version of
SqueezeOS.

A player will make the TimeZone request whenever its own TimeZone has not been
initialized. This would be the case on first set up, or following a factory
reset.

This service was formerly provided by MySqueezeBox.com.

There is one small change in behaviour. MSB would have determined its response
from the location of the player, whereas this implementation will determine
its response from the location of the LMS server.

Signed-off-by: Martin Williams <martinr.williams@gmail.com>

michaelherger commented on Slim/Web/SqueezeosTimezone.pm:
```diff
@@ -0,0 +1,161 @@
+package Slim::Web::SqueezeosTimezone;
```
Please move to just `Slim/Web/Time`. The shorter name should leave options for future developments. You never know.

mw9 replied:
Will do.

Are you happy with the single `/tz` endpoint, or perhaps it might become `/time/tz`. To leave options open.
In the latter case, I would propose changing SqueezeOS accordingly, now. Ideally we wouldn't want to revisit that.

michaelherger replied:
I like the `/time/tz` approach. Thanks!

mw9 replied:
Fixed in latest commit.

michaelherger commented on Slim/Web/SqueezeosTimezone.pm:
```diff
@@ -0,0 +1,161 @@
+	my ($httpClient, $response) = @_;
+
+	# API call to retrieve date/time data
+	Slim::Networking::SimpleAsyncHTTP->new(
```
Please store the timezone in a module global variable and return that value if it's defined. There's no need to look the data up more than once per runtime.

mw9 replied:
Fixed in later commit.

michaelherger commented on Slim/Web/SqueezeosTimezone.pm:
```diff
@@ -0,0 +1,161 @@
+
+	$log->error("TimeZone query: Failed to get TimeZone from ", join("\n", $http->url, $http->error));
+
+	$response->code(RC_NOT_FOUND);
```
I don't think this should be considered a 404 - after all the endpoint has been found. The failure could be manyfold - so rather a 500 (unexpected error). But I don't know about the impact on the caller's side of things.

mw9 replied:
I'll change that -  500 works fine for SqueezeOS, it just logs '500 Internal Server Error'.

mw9 replied:
Fixed in later commit.

michaelherger commented on Slim/Web/SqueezeosTimezone.pm:
```diff
@@ -0,0 +1,161 @@
+		$response->code(RC_OK);
+		$log->info("TimeZone query: Returning \"$tz\" to SqueezeOS device");
+	} else {
+		$response->code(RC_NOT_FOUND);
```
I don't like the 404. Should we fall back to a default timezone instead? If the client can handle it I'd rather return a 204 (successfully executed, but no data), or even a 200 with an empty body.

mw9 replied:
I'll change that to a 204 - SqueezeOS treats it as an error and logs '204 No Content'.

A 200 with empty body doesn't work, we just never see it in the callback. (I had tried that...)

I wouldn't return a default timezone, because SqueezeOS has already set the default `Factory` timezone (GMT), and is using that as the flag to make the request in the first place.

mw9 replied:
Now fixed in later commit.

michaelherger commented on Slim/Web/SqueezeosTimezone.pm:
```diff
@@ -0,0 +1,161 @@
+
+	# All done, return result to SqueezeOS.
+	# But a 404 error if TimeZone failed validation.
+	unless ($tz eq '') {
```
Someone once told me that regular `if` statements were to be preferred over `unless`. After all this could be much simplified to a simple `if ($tz) { ...`.

mw9 replied:
Now fixed.

michaelherger commented on Slim/Web/SqueezeosTimezone.pm:
```diff
@@ -0,0 +1,161 @@
+	#  Some "legacy" and "etc" TimeZones may also contain 0-9 and '+', but
+	#  we do not expect or support such oddities.
+
+	$tz = '' if $tz =~ m{[^A-Za-z._\-/]} ; # reject if any characters outside that range
+
+	# Additional restrictions
+	$tz = '' if $tz =~ m{^/}; # leading '/' not allowed
+	$tz = '' if $tz =~ m{/$}; # trailing '/' not allowed
+	$tz = '' if $tz =~ m{//}; # no path component to be empty
+	$tz = '' if $tz =~ m{ ^- | /- }x; # no path component to start with a hyphen
+
+	# Path components that consist of singleton '.' or doubleton '..' are
+	# not allowed for obvious reasons.
+	# Other than that, the "theory" section of the tz distribution does
+	# allow "dots" elsewhere in TimeZone identifiers, but discourages
+	# them. That said, there are none defined at present, and almost
+	# certainly won't be. So just exclude any TimeZone containing a dot.
+	$tz = '' if $tz =~ m{\.}; # no path component to contain a '.'
+
+	# The 'Factory' TimeZone should never be encountered. SqueezeOS uses
+	# it to signal that the device's TimeZone has not been set.
+	# The 'Etc/Unknown' TimeZone signals an 'Unknown or Invalid' TimeZone,
+	# and is not defined in the tz database. SqueezeOS doesn't want it.
+	$tz = '' if $tz eq 'Factory';
+	$tz = '' if $tz eq 'Etc/Unknown';
```
Please make this a 

```perl
if ($tz =~ /.../) {
} elsif ($tz =~ /.../) {
...
}
```

structure. Once a condition has been met and the `$tz` has been reset, all the other conditions can be ignored.

It could even be a large if with all the conditions `or`ed together:

```perl
if (
	$tz =~ /regex1/ 
	|| $tz =~/regex2/
	|| $tz =~/regex3/ 
	|| ...
) ) {
	$tz = '';
}
```

This way only the minimum number of checks is run.

mw9 replied:
Now cleaned up, as suggested.

michaelherger commented on Slim/Web/SqueezeosTimezone.pm:
```diff
@@ -0,0 +1,161 @@
+	# allow "dots" elsewhere in TimeZone identifiers, but discourages
+	# them. That said, there are none defined at present, and almost
+	# certainly won't be. So just exclude any TimeZone containing a dot.
+	$tz = '' if $tz =~ m{\.}; # no path component to contain a '.'
```
This condition is already covered by the first regex.

Please don't include all the theory. The mentioning of a path got me really confused... And there's a link to the theory at the top anyway.

mw9 replied:
Now dealt with.

michaelherger commented on Slim/Web/SqueezeosTimezone.pm:
```diff
@@ -0,0 +1,161 @@
+	if (!defined $tz || ref $tz) {
+		$log->error('Unexpected JSON response, expected a timezone string: ' . $http->content);
+		$tz = ''; # guarantee that $tz will be a string scalar
+	}
```
Use an `else` here, as we can skip all the rest of the validation in case we didn't have a valid timezone. Or use a shortcut as outlined above in the failure case.

mw9 replied:
I think this calls for a 500  'Internal Server Error' response. And can be routed to a helper function to skip the rest of the validation.

mw9 replied:
Helper function implemented, and case now dealt with as a 500 response.

michaelherger commented on Slim/Web/SqueezeosTimezone.pm:
```diff
@@ -0,0 +1,161 @@
+	my $res = eval { from_json($http->content) };
+	if ($@ || ref $res ne 'HASH') {
+		$log->error($@ || 'Invalid JSON response: ' . $http->content);
+		$res = {}; # guarantee that $res will be a hash ref
```
Maybe you should cut short here, calling `tzAPIError()` or another helper to return the empty result.

mw9 replied:
I am implementing a helper function through which to route all HTTP responses, which will fit in well here.
A 500  'Internal Server Error' response seems appropriate.

mw9 replied:
Now dealt with as a 500 response through a helper function.

michaelherger reviewed (changes requested):
Thanks for this PR! I think that's a much better approach.

Please don't be shocked by the number of comments. I believe the implementation is sound. But it deserves some optimisation. In its current implementation it would run all checks, always. But once we've failed a check, there's no reason to continue the validation. 

And there's no need to look the timezone up on each call. Just do it once. It shouldn't change for the rest of the server's runtime.

mw9 commented:
> Please don't be shocked by the number of comments.

I'm shock-proof. 😄 

A procedural question:

I think the final result should be go forward as one single commit.

Do you prefer to see (a) individual commits that address the points raised prior to reaching that final result, or (b) a succession of "force pushes" ? I never really know what's best.

michaelherger commented:
> Do you prefer to see (a) individual commits that address the points raised prior to reaching that final result, or (b) a succession of "force pushes" ? I never really know what's best.

Thanks for asking! It's a topic we can't agree about on the job 😁. I personally prefer to have the history, thus multiple commits. In particular if I've reviewed a PR already. Otherwise Github will not be able to show me the latest changes etc. With forced pushes and squashes I'd always have to start a review afresh.

That said: I do sometimes squash myself, but try to only do this as long as nobody else has been working on the branch or has reviewed my PR for the above reasons.

mw9 committed:
Provide a default TimeZone to SqueezeOS - Changes following Code review

Renames module from 'Slim/Web/SqueezeosTimezone' to 'Slim/Web/Time'.

Changes HTTP endpoint from '/tz' to '/time/tz'.

Tidies up TimeZone validation checks to minimize the number of checks to be
made, and curtails unnecessary code comments.

Code simplification - sends all HTTP responses through a single helper
function: '_sendHTTPresponse'.

Replaces inappropriate HTTP 404 (Not found) response codes with:
  - where there is an API problem - 500 (Internal server error)
  - where the TimeZone string fails validation - 204 (No content)

Caches the TimeZone obtained from the API call for the duration of LMS's
run time, in a module global variable. This includes TimeZones that have
failed validation (i.e. recorded as ''), because the result is unlikely to
change if the API is called again. But API failures are excluded from this,
on the assumption that these are likely to be transient in nature.

Signed-off-by: Martin Williams <martinr.williams@gmail.com>

michaelherger commented on Slim/Web/Time.pm:
```diff
@@ -63,17 +84,19 @@ sub tzAPIsuccess {
+	# 204 (No content) response.
 
 	my $tz = $res->{'timezone'};
 	if (!defined $tz || ref $tz) {
```
You should remove the `defined` and check for any falsy value: if the returned result was an empty string, we can stop right here.

mw9 replied:
Yes, now done.

mw9 commented:
I have added one commit with changes covering all the points raised. Please let me know if they make sufficient sense.

michaelherger commented on Slim/Web/Time.pm:
```diff
@@ -28,13 +31,29 @@ my $log = logger('network.http');
 sub tzAPIrequest {
 	my ($httpClient, $response) = @_;
 
+	if (defined $cachedAPIresult) {
+		# Did it pass validation ? Or do we have '' ?
+		if ($cachedAPIresult) {
+			$log->info("TimeZone query: Returning \"$cachedAPIresult\" to SqueezeOS device (from cache)");
+			_sendHTTPresponse($httpClient, $response, RC_OK, $cachedAPIresult);
+		}
+		else {
+			$log->error("TimeZone query: Cached timezone was invalid. Returning 204 response to SqueezeOS device");
+			_sendHTTPresponse($httpClient, $response, RC_NO_CONTENT, '');
+		}
+		return;
+	}
```
Minor detail: I'd remove the outer condition, as the first inner condition would imply a defined `$cachedAPIResult`. Therefore this could become just:

```perl
if ($cachedAPIresult) {
	$log->info();
	return _sendHTTPResponse();
}
elsif (defined $cachedAPIResult) {
	$log->warn();
	return _sendHTTPResponse();
}
```

And I'd change from `error()` to `warn()` because otherwise you'd get that message all the time. But the error already gets logged when we receive the value.

mw9 replied:
It may be more "friendly" to simply remove that `elsif`clause altogether. It only gets fired if we have cached the "failed validation" `$tz=''` case. And perhaps we simply shouldn't cache that.

My LMS stays up and running for several weeks or more when I'm not fiddling with it. My guess is that the `$tz=''` case would probably result from a glitchlet in the API, which may be noticed and fixed somewhat sooner.

But I don't feel strongly. Whatever you think is best.

michaelherger replied:
I'm fine with not caching the failure state. Go for it. 

BTW: last night I was working on TIDAL - which for some requests would accept time zone information... I guess I'll extend your work to provide the information internally via a CLI request or similar 😀.

mw9 replied:
Right, I will go with that.

I have also renamed the cache variable from `$cachedAPIresult` to `$cachedTimeZone` to add a little clarity.

Well, a CLI request was how I started out in the CF plugin... 😄.

michaelherger replied:
I still believe it's good to have a web endpoint to simplify changes on the player side. But it should be relatively easy to re-use your code to also provide that information over a CLI request. Thanks a lot for the initiative!

mw9 replied:
I agree, it is much better result.

Thank you for your patient guidance...

michaelherger reviewed (changes requested):
Nice! I only have two minor comments.

mw9 committed:
Provide a default TimeZone to SqueezeOS - Changes following Code review

Rename module variable '$cachedAPIresult' to '$cachedTimeZone' for
clarity.

Only cache a TimeZone string that has passed validation.

Tidy up error detection of an unexpected API response. 


Signed-off-by: Martin Williams <martinr.williams@gmail.com>

mw9 commented:
Further commit added with changes covering off the recent points. They seem to work as expected here.
